actionpack/lib/action_controller/metal/strong_parameters.rb

```diff
@@ -175,15 +175,17 @@ def ==(other_hash)
     #   safe_params.to_h # => {"name"=>"Senjougahara Hitagi"}
     def to_h
       if permitted?
-        @parameters.to_h
+        @parameters.deep_dup
       else
         slice(*self.class.always_permitted_parameters).permit!.to_h
```

Comment:
Not necessary. `.permit!` return `self` which is a `Parameters` object. Calling `to_h` on `Parameters` resulting in calling line 178. (a.k.a. it's getting called by recursion)